src/Storages/StoragePostgreSQL.cpp

```diff
@@ -216,6 +221,63 @@ void StoragePostgreSQL::read(
     query_plan.addStep(std::move(reading));
 }
 
+class SerializationPostgreSQLFixedString : public SerializationFixedString
+{
+private:
+    size_t size;
+
+public:
+    explicit SerializationPostgreSQLFixedString(size_t n_) : SerializationFixedString(n_), size(n_) {}
+
+    void serializeText(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings &) const override
+    {
+        const auto * nested_column = &column;
+        if (column.isNullable())
+        {
+            const auto * column_nullable = assert_cast<const ColumnNullable *>(&column);
+            nested_column = &column_nullable->getNestedColumn();
+        }
+
+        const auto column_type = nested_column->getDataType();
+        const char * data = nullptr;
+        if (isFixedString(column_type))
+        {
+            data = reinterpret_cast<const char *>(
+                &assert_cast<const ColumnFixedString *>(nested_column)->getChars()[size * row_num]);
+        }
+        else if (isString(column_type))
+            data = assert_cast<const ColumnString &>(column).getDataAt(row_num).data;
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't serialize type {} as a FixedString", nested_column->getName());
+
+        /// Write only the amount of UTF-8 characters actually used. PostgreSQL complains if we pass
+        /// any \0 because it doesn't interpret them as valid UTF-8.
+        auto write_size = size;
+        for (size_t i = 0; i < size; ++i)
+        {
+            if (data[i] == 0)
+            {
+                write_size = i;
+                break;
+            }
+        }
+        writeString(data, write_size, ostr);
+    }
+};
+
+static void changeSerializationForPostgreSQLFixedString(const ColumnPtr & column, DataTypePtr & data_type)
+{
+    /// pqxx throws an error when trying to ingest fixed-size strings padded with \0:
+    /// pqxx::data_exception: Failure during '[END COPY]': ERROR:  invalid byte sequence for encoding "UTF8": 0x00
+    auto non_nullable_type = removeNullable(data_type);
+    auto non_nullable_column = removeNullable(column);
+    if (isFixedString(non_nullable_type) && !data_type->getCustomSerialization())
+    {
+        auto n = assert_cast<const DataTypeFixedString &>(*non_nullable_type).getN();
+        auto custom_desc = std::make_unique<DataTypeCustomDesc>(std::make_unique<DataTypeCustomFixedName>(fmt::format("FixedString({})", n)), std::make_shared<SerializationPostgreSQLFixedString>(n));
+        DataTypeFactory::instance().setCustom(data_type, std::move(custom_desc));
```

Comment:
Changing the serialization for this type is the less ugly way and the best abstraction that I found to workaround this. In order to make it work, I had to create a new `setCustom` to in `DataTypeFactory` because it's a private method of `IDataType`